### ISSUE TITLE
Compiled models are measurably worse compared to the parent models from hugging face (without any weight quantization q0f16)

### ISSUE THREAD

**Issue by vin136** [mlc-ai/mlc-llm#1819]
## 🐛 Bug

Mistral7B with `q0f16` compiled via `MLC` does not match the corresponding Huggingface version. Since we can't get the output logits with MLC, I have set the temperature to zero and compared it. Most specifically, after about 2 or 3 turns, further exchanges will either repeat or output hyperlinks, or other unrelated content. 

Here's a way to verify:
My hardware: NVIDIA 4090,CUDA 12.1
Fix system prompts+interaction(sequence of prompts)

For MLC
temperature=0.0
model,weights = Mistral-7B-Instruct-v0.2-q0f16-MLC,Mistral-7B-Instruct-v0.2-q0f16-cuda.so
Note: These are compiled but similar results are observed using the precompiled q4f16_1 ones.
For Hugging-face:
Greedy sampling with float16 weights.

## To Reproduce
Below are the two files, one for running `mlc` model and the other for the corresponding one from Hugging Face model.
[MLC code ](https://drive.google.com/file/d/1jiAwCe4QbpC2hpVc1qFctPl1N3BDFOVH/view?usp=sharing)
[Huggingface Code](https://drive.google.com/file/d/1pJzJFIB9XfQoEpJ8NBXpf2w7ZiY4Dfep/view?usp=sharing)

```from mlc_chat import ChatModule, ChatConfig, ConvConfig,GenerationConfig
from mlc_chat.callback import StreamToStdout
from mlc_chat.interface.openai_api import ChatMessage
from mlc_chat.chat_module import PlaceInPrompt
from threading import Thread
from mlc_chat.callback import StreamIterator
from collections import defaultdict

from pathlib import Path
import pandas as pd
import time
import os
from collections import namedtuple

system_prompt = """You are a **Helpful Voice Assistant**. You should:
1. Provide **Concise** and **Short** responses. 
2. **Do not** ask the user for any further requests,he'll provide if he has one.
3. If a request is beyond your capacity just reply: 'Sorry,I can't answer that'
4. Keep your responses under 2 sentences
5. Do not output images,or hyperlinks, only text that could be read"""


workdir = Path(os.getcwd()).parent


#chat_config = ChatConfig(max_gen_len=256,conv_config=conv_config)

#import pdb;pdb.set_trace()
cm = ChatModule(
    model=str(workdir/"dist/Mistral-7B-Instruct-v0.2-q0f16-MLC"),
    model_lib_path=str(workdir/"dist/libs/Mistral-7B-Instruct-v0.2-q0f16-cuda.so"),
    device='cuda',
    chat_config=ChatConfig(
                max_gen_len=1000,
                conv_config=ConvConfig(system=system_prompt),
                temperature= 0.0
            )
    )
prompts = ["Hey,can you tell me a bad joke?",
               "By the way my name is Luis,what's your name?",
               "Ok,What's the capital of Andhra Pradesh?","Ok,Where's charminar?",
           "What is two plus two?",
           "What was the first joke you told me?",
           "Can you recall my name?"]


if __name__=='__main__':
    state = []
    cm._process_system_prompts()
    for i,prompt in enumerate(prompts):
        print(f"user:{prompt}")
        state.append(ChatMessage(role='user',content=prompt))
        st = time.perf_counter()
        #sampling_params = GenerationConfig(stop=['[','('])
        #sampling_params =None
        out = cm.generate(
                prompt,
                stateless=False
            )
        end = time.perf_counter()
        print(f"MistAI:{out}")
        state.append(ChatMessage(role='assistant',content=out))
        #print(f'*********{end-st}**************')

```

## Expected behavior

Ideally the outputs should nearly match, at least for this simple exchange, responses should be good. 

## Additional context(MLC sample response)

<img width="1340" alt="Screenshot 2024-02-22 at 5 04 00 PM" src="https://github.com/mlc-ai/mlc-llm/assets/21222766/bf382748-1e6b-46cf-9d7b-08be188842c1">

**Comment by vinx13** [comment]
@Hzfengsy  One of the different is PyTorch is we are using f16 accumulator, shall we change it to f32 by default? https://github.com/apache/tvm/blob/main/src/runtime/contrib/cublas/cublas.cc#L168

**Comment by Hzfengsy** [comment]
Not sure if it's because of the acc dtype. 

> after about 2 or 3 turns, further exchanges will either repeat or output hyperlinks, or other unrelated content.

I guess it may be related to the sliding window. also cc @tqchen

**Comment by tqchen** [comment]
@vinx13 let us send a PR to change acc dtype. We are in the process to migrating to new kv, and should dig into this again after migration completed, cc @CharlieFRuan

**Comment by MasterJH5574** [comment]
Meanwhile we are adding `log_prob` support to our serving framework so hopefully that can reflect some logit information when finished.

**Comment by vin136** [comment]
@MasterJH5574  that would be excellent. On that note, would it also allow flexible sampling on the user's end - Say we want to enforce constraints(eg: JSON schema). Essentially use MLC for forward pass(and prefill/cache) and let users control sampling.

**Comment by MasterJH5574** [comment]
Thank you @vin136 for following up! We just rolled the logprob support out in #1832.

> Say we want to enforce constraints(eg: JSON schema).

Yes we are also working on JSON schema @Ubospica is leading this. #1534, #1801 are the PRs that bring in grammar support. The JSON schema support will come shortly.

> Essentially use MLC for forward pass(and prefill/cache) and let users control sampling.

I agree with you that logits can be customized by users prior to sampling. For the first step, we are going to bring the logit masking support (which backs the JSON schema support). We will try to extend the customizability, but that might not be an immediate work.

**Comment by vin136** [comment]
@MasterJH5574  First thanks for pointing out the latest developments. I want to bring to your notice some minor syntactical bugs(could be major,I'm not sure!) in the current main branch. Here is the issue #1859 . I'm new to TVM but trying to debug it at the moment. Thanks!
